epi/epi_functions.py

```diff
@@ -0,0 +1,265 @@
+
+    try:
+        xy = pd.read_excel(
+            xlsx,
+            header=0,
+            sheet_name="tb4",
+            skiprows=[0, 1],
+            parse_dates=True,
+            names=custom_column_names,
+            usecols=custom_column_indices,
+            converters=dict(
+                (i, int) for i in range(len(custom_column_indices)) if i > 0
+            ),
+            skipfooter=1,  # skips the SKUPAJ row at the end
+        )
+        xy["datum"] = pd.to_datetime(xy["datum"])
+    except XLRDError as xe:
+        exit(
+            f"Unable to find sheet named tb4. Please examine {xlsx} if the sheet name "
+            f"has been spelled correctly."
+        )
+
+    # 2: data integrity check
+    #   - expect first column to be date and monotinically increasing/decreasing
+    if not xy[xy["datum"].diff().dt.days < 0].empty:
+        raise ValueError(
+            "Datums are monotonically increasing. Please check the datum column to "
+            "see if dates are in order."
+        )
+
+    #   - expect columns 2:11 to be counts for age groups 0-85+ for men
+    #   - expect columns 13:22 to be counts for age groups 0-85+ for women
+    # This is resolved on import using `converters`.
+
+    #   - all values for columns 2:22 should be >= 0 or NA (empty)
+    expect_counts = xy.iloc[:, 1 : len(xy.columns)]
+    expect_counts_gezero = expect_counts >= 0
+    expect_counts_null = expect_counts.isnull()
+    counts_or_nan = expect_counts_gezero.__or__(expect_counts_null)
+
+    bad_indices = xy.loc[~counts_or_nan.all(axis=1)]
+    if len(bad_indices) > 0:
+        raise ValueError(
+            f"Some counts are not greater than 0 or are not NaN. Hint:\n\n"
+            f"{bad_indices}"
+        )
+
+    xyall = pd.DataFrame(
+        {"datum": pd.date_range(start=xy["datum"].min(), end=xy["datum"].max())}
+    )
+
+    # 3: missing dates
+    #   - if any missing dates, ask user if assuming NA is ok
+    #     - if ok, add missing dates NA for all columns
+    #     - if not, abort
+    if xyall.shape[0] != xy.shape[0] and impute:
+        # pr_datum = input("Some dates are missing. Impute with NaN? (yes/no): ")
+        # if pr_datum in ["y", "yes"]:
+        xy = xyall.merge(xy, how="left", on="datum")
+
+    # Add 0 to missing fieds in order for the cumsum function to output expected result.
+    # If this is not done, the missing values create a gap in cumulated sum.
+    xy = xy.fillna(0)
+
+    # 4: calculate sums
+    #   - calculate cumulative sums per age group per gender (20 new variables)
+    cs = xy.iloc[:, 1:len(xy.columns)].cumsum(axis=None)
+    #   - calculate row-wise age group sums (irrespective of gender) of cumulative
+    #   sums (10 new variables)
+    age_groups = [
+        "0-4",
+        "5-14",
+        "15-24",
+        "25-34",
+        "35-44",
+        "45-54",
+        "55-64",
+        "65-74",
+        "75-84",
+        "85+",
+    ]
+    rwgs = []  # row-wise group sums
```

Comment:
Maybe make this name slightly less cryptic for anyone not reading this particular comment? The cvs file name seems to be the biggest victim.